From a0a0a8ab967b72be4a8e7ed171a342ca5c0dcf0a Mon Sep 17 00:00:00 2001
From: guanwenx <guanx.wen@intel.com>
Date: Thu, 4 Dec 2014 17:41:48 +0800
Subject: [PATCH 165/176] ov5693: fix the exposure and gain setting.

In the past, gain effect in next frame, but exposure time effect in
next 2 frame.After this fix, both gain and exposure will effect in
2 frame.

Change-Id: I8b0a22f176ecf5c6a6eb41cbc7b8e5eb769ab66a
Tracked-On: https://jira01.devtools.intel.com/i#browse/GMIN-3597
Signed-off-by: Guan Wen <guanx.wen@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/31423
Reviewed-by: Bin B Han <bin.b.han@intel.com>
Reviewed-by: Jing Tao <jing.tao@intel.com>
Reviewed-by: Jukka O Kaartinen <jukka.o.kaartinen@intel.com>
Approver: Wei Tao <wei.tao@intel.com>
Build: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
Maintainer: Wei Tao <wei.tao@intel.com>
---
 .../camera/drivers/media/i2c/ov5693-ecs/ov5693.c   | 43 ++++++++--------------
 .../camera/drivers/media/i2c/ov5693-ecs/ov5693.h   |  2 +-
 2 files changed, 16 insertions(+), 29 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c b/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
index ed13182..7ef0669 100755
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
@@ -602,49 +602,36 @@ static long __ov5693_set_exposure(struct v4l2_subdev *sd, int coarse_itg,
 		return ret;
 	}
 
-	ret = ov5693_write_reg(client, OV5693_8BIT, OV5693_AGC_H, (gain >> 8) & 0xff);
-	if (ret) {
-		dev_err(&client->dev, "%s: write %x error, aborted\n",
-			__func__, OV5693_AGC_H);
-		return ret;
-	}
-
 	/* Digital gain */
 	if (digitgain) {
-		ret = ov5693_write_reg(client, OV5693_16BIT,
-				OV5693_MWB_RED_GAIN_H, digitgain);
-		if (ret) {
-			dev_err(&client->dev, "%s: write %x error, aborted\n",
-				__func__, OV5693_MWB_RED_GAIN_H);
-			return ret;
-		}
-
-		ret = ov5693_write_reg(client, OV5693_16BIT,
-				OV5693_MWB_GREEN_GAIN_H, digitgain);
-		if (ret) {
-			dev_err(&client->dev, "%s: write %x error, aborted\n",
-				__func__, OV5693_MWB_RED_GAIN_H);
-			return ret;
-		}
-
-		ret = ov5693_write_reg(client, OV5693_16BIT,
-				OV5693_MWB_BLUE_GAIN_H, digitgain);
+		int setvalue = 0;
+
+		/*
+		*  for sensor metadata only support 1x/2x digital gain, to support
+		*  sensor metadata we need switch from MWB gain to digital gain
+		*/
+		if (digitgain == 2048)
+			setvalue = 1;
+		else
+			setvalue = 0;
+
+		ret = ov5693_write_reg(client, OV5693_8BIT, OV5693_AGC_H, setvalue);
 		if (ret) {
 			dev_err(&client->dev, "%s: write %x error, aborted\n",
-				__func__, OV5693_MWB_RED_GAIN_H);
+				__func__, OV5693_AGC_H);
 			return ret;
 		}
 	}
 
 	/* End group */
 	ret = ov5693_write_reg(client, OV5693_8BIT,
-			       OV5693_GROUP_ACCESS, 0x10);
+				OV5693_GROUP_ACCESS, 0x10);
 	if (ret)
 		return ret;
 
 	/* Delay launch group */
 	ret = ov5693_write_reg(client, OV5693_8BIT,
-					   OV5693_GROUP_ACCESS, 0xa0);
+				OV5693_GROUP_ACCESS, 0xa0);
 	if (ret)
 		return ret;
 	return ret;
diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.h b/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.h
index d9983b8..dfee36a 100755
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.h
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.h
@@ -317,7 +317,7 @@ static struct ov5693_reg const ov5693_global_setting[] = {
 	{OV5693_8BIT, 0x3405, 0x00},
 	{OV5693_8BIT, 0x3406, 0x01},
 	{OV5693_8BIT, 0x3500, 0x00},
-	{OV5693_8BIT, 0x3503, 0x07},
+	{OV5693_8BIT, 0x3503, 0x17},
 	{OV5693_8BIT, 0x3504, 0x00},
 	{OV5693_8BIT, 0x3505, 0x00},
 	{OV5693_8BIT, 0x3506, 0x00},
-- 
1.9.1

